File: instagram/src/main/java/main/entity/Post.java
package main.entity;

import jakarta.persistence.*;
import lombok.Data;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Data
@Entity
@Table(name = "posts")
public class Post {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String text;

    @Column(name = "image_path")
    private String imagePath;

    @Enumerated(EnumType.STRING)
    private PostStatus status;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User author;

    @ManyToOne
    @JoinColumn(name = "parent_id")
    private Post parent;

    @OneToMany(mappedBy = "parent", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Post> posts = new ArrayList<>();

    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Vote> votes = new ArrayList<>();

    @ManyToMany
    @JoinTable(name = "post_tags",
            joinColumns = @JoinColumn(name = "post_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id"))
    private Set<Tag> tags = new HashSet<>();

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
        this.status = PostStatus.IN_PROGRESS;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}

File: frontend/app/entities/post.ts
import type { User } from "./user";
import type { Tag } from "./tag";
import type { Vote } from "./vote";

export interface Post {
    type: string;
    id: number;
    attributes: PostAttributes;
    relationships: PostRelationships;
    links: PostLinks;
    votes: Vote[];
    tags: Tag[];
}

export interface PostAttributes {
    title: string;
    text: string;
    status: "NEW" | "ACTIVE" | "FIRST_REACTION" | "OUTDATED";
    imagePath: string;
    createdAt: string;
    updatedAt: string;
    tags: string[];
    is_commentable: boolean;
}

export interface PostRelationships {
    author: User|null;
    comments: Post[];
    post: Post;
}

export interface PostLinks {
    self: string;
    parent: string;
}

